### PR TITLE
Add support for filtering uncategorized tests with TestCategory=None

### PR DESCRIPTION
**Pull request by Evangelink**
## Summary

Add support for filtering uncategorized tests using the reserved keyword `None` (e.g. `TestCategory=None`). This enables users to select or exclude tests that have no category assigned using standard filter syntax.

Inspired by [microsoft/testfx#6779](https://github.com/microsoft/testfx/pull/6779) and the design discussion in [microsoft/testfx#5136](https://github.com/microsoft/testfx/issues/5136).

## Motivation

Users want to find/run tests that haven't been categorized yet (e.g. to track categorization progress). Currently there is no way to filter for tests without a `TestCategory` using `--filter`. The agreed design uses `None` as a reserved keyword meaning "uncategorized".

### `Condition.cs`
- Added `NoneFilterValue` constant (`"None"`)
- **EvaluateEqualOperation**: When `Value` equals `"None"` (case-insensitive) and property is `null` or empty array, returns `true` (uncategorized match). `NotEqual` gets the inverse behavior automatically.

### `FastFilter.cs`
- **Evaluate**: Check if filter set contains `"None"`. When property provider returns `null` or empty array and `"None"` is in the filter set, treat as matched.

## Filter semantics

| Filter | Matches |
|--------|---------|
| `TestCategory=None` | Tests with no category (null/empty property) |
| `TestCategory!=None` | Tests that have at least one category |
| `TestCategory=None\|TestCategory=CategoryA` | Uncategorized OR CategoryA |

## Tests

- **Unit tests** (ConditionTests.cs): 9 new tests covering parsing and evaluation with `None` for Equal/NotEqual, case-insensitivity, and null/empty/non-empty properties
- **Unit tests** (FastFilterTests.cs): 3 new tests for FastFilter with `None` equal, equal+OR, and not-equal
- **Acceptance tests** (TestCaseFilterTests.cs): 2 new end-to-end tests validating `TestCategory=None` and `TestCategory!=None` against the SimpleTestProject sample

### CODE REVIEW

File: src/Microsoft.TestPlatform.Filter.Source/FastFilter.cs [hunk 1]
```diff
@@ -93,11 +102,16 @@ internal bool Evaluate(Func<string, object?> propertyValueProvider)
...
+            else if (multiValues is { Length: > 0 })
             {
                 var values = PropertyValueRegex == null ? multiValues : multiValues?.Select(value => ApplyRegex(value));
                 matched = values?.Any(result => result != null && FilterProperties[name].Contains(result)) == true;
             }
+            else if (hasEmptyStringFilter)
+            {
+                // Empty array matches empty string filter.
+                matched = true;
+            }
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/Microsoft.TestPlatform.Filter.Source/FastFilter.cs]: Good catch. The inconsistency was that FastFilter's `Contains(string.Empty)` on the filter set naturally matched when a property value was `""`, but Condition's early return only checked for `null`/empty arrays and never let `[""]` reach the normal equality loop.

Fixed in 4c91040:
- `EvaluateEqualOperation`: Changed to only special-case `null`/empty arrays in the early return. Arrays with elements (including `[""]`) now fall through to the normal `string.Equals` loop, which naturally matches `"" == ""` — consistent with FastFilter's `Contains("")` behavior.
- `EvaluateContainsOperation`: Changed to use `Array.TrueForAll(multiValue, static v => v.Length == 0)` so that arrays of empty strings are also treated as uncategorized. (We still need the early return here because `string.IndexOf("")` always returns 0, which would match all tests.)

Added tests for single empty-string property and array-with-empty-string scenarios to verify consistency across both paths.

File: src/Microsoft.TestPlatform.Filter.Source/Condition.cs [hunk 2]
```diff
@@ -120,6 +126,12 @@ private bool EvaluateEqualOperation(string[]? multiValue)
 
     private bool EvaluateContainsOperation(string[]? multiValue)
     {
+        // Special case: empty string filter value matches null/empty property (uncategorized tests).
+        if (Value.Length == 0)
+        {
+            return multiValue is null or { Length: 0 };
+        }
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/Microsoft.TestPlatform.Filter.Source/Condition.cs]: Correct — `GetPropertyValue` wraps a single `string.Empty` into `[""]`, which had `Length == 1` and failed the `is null or { Length: 0 }` check.

Fixed in 4c91040:
- `EvaluateEqualOperation`: The early return now only handles `null`/empty arrays. A `[""]` array falls through to the normal `string.Equals` loop where `"" == ""` matches — consistent with FastFilter.
- `EvaluateContainsOperation`: Now uses `Array.TrueForAll(multiValue, static v => v.Length == 0)` which returns `true` for both empty arrays and arrays of empty strings.

Added 5 new tests specifically covering empty-string property values (single string and array) to verify the paths produce the same results.

**Comment by nohwnd** [comment]
In the linked issue we agreed on implementing this as `TestCategory=None` ("reserving" None for our use as it probably is rarely used), rather than `TestCategory=`.

This PR seems to implement the second approach instead.

https://github.com/microsoft/testfx/issues/5136#issuecomment-3450735963

**Comment by Evangelink** [comment]
Thanks for the pointer! Updated the implementation to use `TestCategory=None` as agreed in the design discussion. The reserved keyword approach is cleaner — no parsing changes needed, just evaluation logic.

Summary of changes in 1be059a:
- Reverted all Parse changes — empty values are rejected again as before
- Added `NoneFilterValue = "None"` constant to `Condition`
- `EvaluateEqualOperation`: when Value equals "None" (case-insensitive) and property is null/empty, returns true
- `FastFilter.Evaluate`: checks for "None" in filter set instead of empty string
- Updated all unit and acceptance tests to use `TestCategory=None` / `TestCategory!=None`

The `NotEqual` and `Contains`/`NotContains` operators work naturally:
- `TestCategory!=None` → matches only categorized tests (via `!EvaluateEqualOperation`)
- `TestCategory~None` / `TestCategory!~None` → regular substring matching (no special handling, "None" is not special for Contains)

**Comment by nohwnd** [comment]
I don't think the tests cover the situation when user explicitly specified "None" as the category (or other value, we now genericly consider None to be absence of value). Should that throw or at least warn? I think it is reasonable if it warns, and inludes the item with None, (and excludes on negative filter).

**Comment by Evangelink** [comment]
Good point — added tests in d0e2bea covering the explicit `[TestCategory("None")]` scenario:

- `EvaluateNoneEqualWithExplicitNoneCategoryShouldReturnTrue` — `TestCategory=None` matches a test with literal "None" category ✓
- `EvaluateNoneNotEqualWithExplicitNoneCategoryShouldReturnFalse` — `TestCategory!=None` does NOT match it ✓
- `EvaluateNoneEqualWithExplicitNoneCategoryAmongOthersShouldReturnTrue` — multi-category `["None", "CategoryA"]` still matches ✓
- `FastFilterWithNoneEqualShouldAlsoMatchExplicitNoneCategory` — FastFilter path behaves the same ✓

The current behavior is: `TestCategory=None` matches both uncategorized tests AND tests with a literal "None" category. This seems right per your comment — include the item with None, exclude on negative filter.

Regarding warnings for users who explicitly use `[TestCategory("None")]` — that would probably be best as an MSTest analyzer (similar to what was mentioned in [testfx#5136](https://github.com/microsoft/testfx/issues/5136#issuecomment-2457539139)). We could flag it as a follow-up issue since it's in testfx's domain rather than vstest's. What do you think?

**Comment by nohwnd** [comment]
thanks for the pr will have a look as soon as I am done fighting with system.collections.immutable update

**Review by nohwnd — APPROVED** [review]

**Comment by nohwnd** [comment]
if you want it released sooner backport to rel/18.7 please

**Comment by nohwnd** [comment]
main is already18.8 now because of move to stj

**Review by Youssef1313 — COMMENTED** [review]
This looks like a breaking change to me? "None" was never previously a reserved keyword and maybe someone is using "None" already for some reason?

**Comment by nohwnd** [comment]
Technically yes, None will still be included in the run, when None filter is used, together with all untagged tests. So we are broadening the scope of tests that will run, so not a severe breaking change where tests would silently not run. Will document it in release notes.
